src/Debug/TraceableDomainEventDispatcher.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Biig\Component\Domain\Debug;
+
+use Biig\Component\Domain\Event\DelayedListener;
+use Biig\Component\Domain\Event\DomainEvent;
+use Biig\Component\Domain\Event\DomainEventDispatcher;
+use Biig\Component\Domain\Event\DomainEventDispatcherInterface;
+use Biig\Component\Domain\Exception\InvalidArgumentException;
+use Biig\Component\Domain\Model\ModelInterface;
+use Biig\Component\Domain\Rule\DomainRuleInterface;
+use Biig\Component\Domain\Rule\PostPersistDomainRuleInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher;
+use Symfony\Component\EventDispatcher\Debug\WrappedListener;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class TraceableDomainEventDispatcher extends TraceableEventDispatcher implements DomainEventDispatcherInterface
+{
+    /**
+     * @var array
+     */
+    private $eventsFired;
+
+    /**
+     * @var DomainEventDispatcher
+     */
+    private $decorated;
+
+    /** @var array */
+    private $delayedListenersCalled;
+
+    /**
+     * TraceableDomainEventDispatcher constructor.
+     *
+     * @param DomainEventDispatcher $dispatcher
+     */
+    public function __construct(DomainEventDispatcher $dispatcher)
+    {
+        $this->eventsFired = [];
+        $this->delayedListenersCalled = [];
+        $this->decorated = $dispatcher;
+        parent::__construct($dispatcher, new Stopwatch(), new NullLogger());
+    }
+
+    public function addDomainRule(DomainRuleInterface $rule)
+    {
+        return $this->decorated->addDomainRule($rule);
+    }
+
+    public function addPostPersistDomainRuleInterface(PostPersistDomainRuleInterface $rule)
+    {
+        return $this->decorated->addPostPersistDomainRuleInterface($rule);
+    }
+
+    public function persistModel(ModelInterface $model)
+    {
+        /** @var DelayedListener $listener */
+        foreach ($this->decorated->getDelayedListeners() as $listener) {
+            $eventName = $listener->getEventName();
+
+            if ($listener->shouldOccur($model)) {
+                if (!$listener instanceof WrappedListener) {
+                    $listener = new WrappedDelayedListener($listener);
+                }
+
+                $this->delayedListenersCalled[] = $listener->getInfo($eventName);
+            }
+        }
+
+        return $this->decorated->persistModel($model);
+    }
+
+    /**
+     * Dispatches an event to all registered listeners.
+     *
+     * For BC with Symfony 4, the $eventName argument is not declared explicitly on the
```

Comment:
Remove this comment. We do not have this problem.

Reply:
OK